content/guides/destructuring.adoc

```diff
@@ -0,0 +1,534 @@
+----
+
+When defining a function in clojure, there is an implicit `let` binding used on the parameters in order to create the local scope of the function. We can leverage this to whittle our function definition down even more.
+
+[source,clojure]
+----
+(defn print-coordinates-3 [[x y z]]
+  (println "x:" x ", y:" y ", z:" z))
+----
+
+That cut the definition down to be less than half of what it was to get the same output.
+
+[source,clojure]
+----
+(print-coordinates-1 [10 50 30])
+;= "x: 10, y: 50, z: 30"
+
+(print-coordinates-2 [10 50 30])
+;= "x: 10, y: 50, z: 30"
+
+(print-coordinates-3 [10 50 30])
+;= "x: 10, y: 50, z: 30"
+----
+
+For a more realistic example, let's create a map containing some basic contact information for the infamous John Smith.
+
+[source,clojure]
+----
+(def john-smith {:f-name "John"
+                 :l-name "Smith"
+                 :phone "2150041776"
+                 :company "Functional Industries"
+                 :title "Sith Lord of Git"})
+----
+
+So here we have John's personal information and now we need to access the values within this more-than-trivial map.
+
+[source,clojure]
+----
+(defn print-contact-info [{:keys [f-name l-name phone company title]}]
+  (println f-name l-name "is the" title "at" company)
+  (println "You can reach him at" phone))
+
+(print-contact-info john-smith)
+;= John Smith is the Sith Lord of Git at Functional Industries
+;= You can reach him at 2150041776
+----
+
+This function will associatively destructure the the input using the `:keys` shortcut and then print out the contact information that we provided.
+
+But what about when we want to send John a nice letter?
+
+[source,clojure]
+----
+(def john-smith {:f-name "John"
+                 :l-name "Smith"
+                 :phone "2150041776"
+                 :address {:street "452 Lisp Ln."
+                           :city "Macroville"
+                           :state "Kentucky"
+                           :zip "81321"}
+                 :hobbies ["running" "hiking" "basketball"]
+                 :company "Functional Industries"
+                 :title "Sith Lord of Git"})
+----
+
+We have an address in there now, but we needed to nest a map into our original structure in order to accomplish this.
+
+[source,clojure]
+----
+(defn print-contact-info [{:keys [f-name l-name phone company title]
+                          {:keys [street city state zip hobbies]} :address}
+                          [fav-hobby second-hobby] hobbies]
+  (println f-name l-name "is the" title "at" company)
+  (println "You can reach him at" phone)
+  (println "He lives at" street city state zip)
+  (println "Maybe you can write to him about" fav-hobby "or" second-hobby))
+
+(print-contact-info john-smith)
+;= John Smith is the Sith Lord of Git at Functional Industries
+;= You can reach him at 2150041776
+;= He lives at 452 Lisp Ln. Macroville Kentucky 81321
+;= Maybe you can write to him about running or hiking
+----
```

Comment:
Alright, here we go! I tried to make the code examples a little more realistic and in depth. I also added a little bit of sequential destructuring to the final example for the sake of having an all-inclusive final showing. Let me know what you think!